### ISSUE TITLE
Regarding CLIP Vision Backbone

### ISSUE THREAD

**Issue by HashmatShadab** [speedlab-git/SimCLIP#1]
Hi!

Thanks for sharing your work! I would like to know that  whether the adversarial results reported for the Llava-1.5 are using CLIP encoder ViT-L/14 or ViT-L/14@336? Does the adversarial evaluation code handle both Vision encoders?

**Comment by hossainzarif** [comment]
Hi, 

In our experiments, we used LLaVA 1.5 with the ViT-L/14 encoder. Our adversarial fine-tuning approach is built upon this version of the CLIP encoder to enhance its robustness.

**Comment by HashmatShadab** [comment]
Thank you. Does the adversarial evaluation code handle both encoders?

**Comment by HashmatShadab** [comment]
As stated in the paper, ViT-L/14 is used as the CLIP encoder and integrated into the LLaVA-1.5 model, while maintaining the pretrained LLM and projection weights unchanged. However, I am slightly confused because the LLaVA-1.5 model uses ViT-L/14@336 as its vision encoder, and the corresponding projection layers were trained during pretraining and instruction tuning with ViT-L/14@336 as vision backbone. So, the image size should be 336x336 and robust encoder should be aligned with ViT-L/14@336?

**Comment by HashmatShadab** [comment]
https://huggingface.co/liuhaotian/llava-v1.5-7b/blob/main/config.json

As can be seen from the above config file, openai/clip-vit-large-patch14-336 is used as the vision encoder for llava-v1.5-7b.

**Comment by hossainzarif** [comment]
Yes, we used the default ViT-Large-Patch14-336 as our base encoder for adversarial fine-tuning. In your previous question, I thought you were asking whether we used ViT-L/14 or ViT-L/32 as the vision encoder.

I hope this answer clarifies your question.

**Comment by HashmatShadab** [comment]
vision_tower = model.get_vision_tower()
        # vision_tower.set_device(device)
        non_llava = True if pretrained_rob_path not in [None, 'None', 'none'] else False
        if not vision_tower.is_loaded:
            vision_tower.load_model(non_llava, pretrained_rob_path)#.to(device=device)

Given that the pretrained_rob_path is passed to load the SimCLIP weights into the vision encoder, that means the non_llava variable is True. So the follwoing lines are used to load the model

        if non_llava:
            import open_clip
            print("using open_clip")
            model_orig, _, image_processor = open_clip.create_model_and_transforms('ViT-L-14', pretrained='openai')
            vision_model = model_orig.visual
            if pretrained_ckpt != 'openai':
                vision_model.load_state_dict(torch.load(pretrained_ckpt, map_location='cpu'))
            # self.image_processor = image_processor
            # self.image_processor = CLIPImageProcessor.from_pretrained(self.vision_tower_name)
            self.image_processor = CLIPImageProcessor.from_pretrained('openai/clip-vit-large-patch14')  # 224
            # model_orig = vision_model
            # llava operates on the second to last layer output, so we remove the last layer
            vision_model.transformer.resblocks = vision_model.transformer.resblocks[:-1]; print("removing last layer of vision model")
            model_orig = ClipVisionModel(
                model=vision_model,
                normalize=lambda x: x,  # images have to be normalized, e.g. as handled by the llava model wrapper
                all_tokens=True, proj=False
            )
            self.vision_tower = model_orig        
            self.vision_tower.device = device

So does that not mean ViT-L/14 is used instead of ViT-L/14@336 ? Let me know if I am missing something

**Comment by HashmatShadab** [comment]
??

**Comment by speedlab-git** [comment]
Apologies for the delay in my response. I realize my last message may not have fully clarified your question and could have added some confusion.

The config file you referenced is indeed from the LLava codebase and is used during model loading for fine-tuning and inference. In our codebase, the load_model() function has been customized to load a fine-tuned encoder, while open_clip is used specifically to load the CLIP model. If you refer to the open-clip documentation https://pypi.org/project/open-clip-torch/, you'll see that the ViT-L-14 model loads a 224px encoder by default.

Additionally, we load a checkpoint for a robust encoder when rob_path is provided. The rob_path can point to either the OpenAI encoder or a custom robust encoder checkpoint. To summarize: if you use the default loader function from the LLava codebase rather than our customized load_model() function, it will load the default configuration from config.json (336px). However, using our customized loader ensures that a 224px encoder is consistently used for all inferences.

I hope this clears things up! Please feel free to reach out if you have any further questions

**Comment by speedlab-git** [comment]
def load_model(self, device_map=None):
        if self.is_loaded:
            print('{} is already loaded, `load_model` called again, skipping.'.format(self.vision_tower_name))
            return

        self.image_processor = CLIPImageProcessor.from_pretrained(self.vision_tower_name)
        self.vision_tower = CLIPVisionModel.from_pretrained(self.vision_tower_name, device_map=device_map)
        self.vision_tower.requires_grad_(False)

        self.is_loaded = True
        
      Here's the code snippet from the haotian-liu/LLaVA codebase for your reference.

**Comment by HashmatShadab** [comment]
> ```
>     vision_tower = model.get_vision_tower()
>     # vision_tower.set_device(device)
>     non_llava = True if pretrained_rob_path not in [None, 'None', 'none'] else False
>     if not vision_tower.is_loaded:
>         vision_tower.load_model(non_llava, pretrained_rob_path)#.to(device=device)
> ```
> 
> Given that the pretrained_rob_path is passed to load the SimCLIP weights into the vision encoder, that means the non_llava variable is True. So the follwoing lines are used to load the model
> 
> ```
>     if non_llava:
>         import open_clip
>         print("using open_clip")
>         model_orig, _, image_processor = open_clip.create_model_and_transforms('ViT-L-14', pretrained='openai')
>         vision_model = model_orig.visual
>         if pretrained_ckpt != 'openai':
>             vision_model.load_state_dict(torch.load(pretrained_ckpt, map_location='cpu'))
>         # self.image_processor = image_processor
>         # self.image_processor = CLIPImageProcessor.from_pretrained(self.vision_tower_name)
>         self.image_processor = CLIPImageProcessor.from_pretrained('openai/clip-vit-large-patch14')  # 224
>         # model_orig = vision_model
>         # llava operates on the second to last layer output, so we remove the last layer
>         vision_model.transformer.resblocks = vision_model.transformer.resblocks[:-1]; print("removing last layer of vision model")
>         model_orig = ClipVisionModel(
>             model=vision_model,
>             normalize=lambda x: x,  # images have to be normalized, e.g. as handled by the llava model wrapper
>             all_tokens=True, proj=False
>         )
>         self.vision_tower = model_orig        
>         self.vision_tower.device = device
> ```
> 
> So does that not mean ViT-L/14 is used instead of ViT-L/14@336 ? Let me know if I am missing something

That's what I was trying to know.  So you have adversarially trained CLIP at 224 only
